PROJECT_MCPU/MCPU/DOC/GENERATED/man/man3/CanDeviceProtocol.3
.TH "CanDeviceProtocol" 3 "MCPU" \" -*- nroff -*-
.ad l
.nh
.SH NAME
CanDeviceProtocol
.SH SYNOPSIS
.br
.PP
.PP
\fR#include <CanDeviceProtocol\&.h>\fP
.PP
Inherited by \fBPCB301\fP, \fBPCB302\fP, \fBPCB303\fP, \fBPCB304\fP, \fBPCB325\fP, and \fBPCB326\fP\&.
.SS "Classes"

.in +1c
.ti -1c
.RI "class \fBCanDeviceCommand\fP"
.br
.ti -1c
.RI "class \fBCanDeviceCommandResult\fP"
.br
.RI "This class provides a data structure to handle a command result\&. "
.ti -1c
.RI "class \fBCanDeviceCommunicationMonitor\fP"
.br
.RI "This class provides an interface to monitor the communication performance with the device\&. "
.ti -1c
.RI "class \fBCanDeviceRegister\fP"
.br
.RI "This class provides a data structure for the can device frame communication protocol\&. "
.ti -1c
.RI "class \fBRegister\fP"
.br
.RI "This is the data structure handling the protocol device register data\&. "
.in -1c
.SS "Public Types"

.in +1c
.ti -1c
.RI "enum class \fBCommandRegisterErrors\fP { \fBCommandRegisterErrors::COMMAND_NO_ERROR\fP = 0, \fBCommandRegisterErrors::COMMAND_ERROR_BUSY\fP, \fBCommandRegisterErrors::COMMAND_ERROR_INVALID_PARAM\fP, \fBCommandRegisterErrors::COMMAND_ERROR_MOMENTARY_DISABLED\fP, \fBCommandRegisterErrors::COMMAND_INVALID_DEVICE\fP =253, \fBCommandRegisterErrors::COMMAND_COMMUNICATION_ERROR\fP =254, \fBCommandRegisterErrors::COMMAND_DEVICE_TMO\fP = 255 }"
.br
.RI "This is the enumeration of the possible command executed errors\&. "
.ti -1c
.RI "enum class \fBstatus_options\fP { \fBstatus_options::WAITING_CAN_DRIVER_CONNECTION\fP = 0, \fBstatus_options::WAITING_REVISION\fP, \fBstatus_options::DEVICE_CONFIGURATION\fP, \fBstatus_options::DEVICE_RUNNING\fP, \fBstatus_options::DEVICE_SIMULATOR\fP, \fBstatus_options::LEN\fP, \fBstatus_options::UNDEF\fP = LEN }"
.br
.RI "This is the enumeration of the Module's internal operating status\&.
.br
See the \fBgetModuleStatus()\fP method\&. "
.ti -1c
.RI "enum class \fBbootloader_options\fP { \fBbootloader_options::BOOTLOADER_NOT_PRESENT\fP = 0, \fBbootloader_options::BOOTLOADER_RUNNING\fP = 1, \fBbootloader_options::BOOTLOADER_PRESENT\fP = 2, \fBbootloader_options::BOOTLOADER_UNCKNOWN_STAT\fP }"
.br
.RI "This is the enumeration calss defining the current status of the remote bootloader activity\&. "
.in -1c
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBCanDeviceProtocol\fP (unsigned char devid, LPCWSTR devname)"
.br
.ti -1c
.RI "void \fBrunMode\fP (void)"
.br
.RI "This function activate the Module in Real mode\&. "
.ti -1c
.RI "void \fBsimulMode\fP (void)"
.br
.RI "This function activate the Module in Simulation mode\&. "
.ti -1c
.RI "bool \fBisSimulatorMode\fP (void)"
.br
.RI "This function tests if the module is running in simulation mode\&. "
.ti -1c
.RI "bool \fBisCommunicationError\fP (void)"
.br
.RI "This function tests if the module is in communication error condition\&. "
.ti -1c
.RI "\fBstatus_options\fP \fBgetModuleStatus\fP (void)"
.br
.RI "This function returns the Module operating status\&. "
.ti -1c
.RI "\fBbootloader_options\fP \fBgetBootStatus\fP (void)"
.br
.RI "This function returns the current status of the bootloader activities on the remote device\&. "
.ti -1c
.RI "System::String ^ \fBgetBootRevision\fP (void)"
.br
.RI "This function returns a string with the current detected Bootloader revision code\&. "
.ti -1c
.RI "System::String ^ \fBgetAppRevision\fP (void)"
.br
.RI "This function returns a string with the current detected remote firmware revision code\&. "
.ti -1c
.RI "bool \fBcommandNoWaitCompletion\fP (unsigned char code, unsigned char d0, unsigned char d1, unsigned char d2, unsigned char d3, int tmo)"
.br
.RI "This is a non blocking thread safe function sending a command to the remote device\&. "
.ti -1c
.RI "bool \fBcommandNoWaitCompletion\fP (\fBCanDeviceCommand\fP^ command, int tmo)"
.br
.ti -1c
.RI "\fBCanDeviceCommandResult\fP ^ \fBcommandWaitCompletion\fP (unsigned char code, unsigned char d0, unsigned char d1, unsigned char d2, unsigned char d3, int tmo, Object^ src)"
.br
.RI "This is a blocking thread safe function sending a command to the remote device\&. "
.ti -1c
.RI "\fBCanDeviceCommandResult\fP ^ \fBcommandWaitCompletion\fP (\fBCanDeviceCommand\fP^ command, int tmo, Object^ src)"
.br
.ti -1c
.RI "bool \fBisCommandCompleted\fP (void)"
.br
.RI "This function returns the current status of the pending command\&. "
.ti -1c
.RI "bool \fBisCommandError\fP (void)"
.br
.RI "This function returns true if the command is terminated with an error condition\&. "
.ti -1c
.RI "unsigned char \fBgetCommandResult0\fP (void)"
.br
.RI "In case of command completed without errors, this function returns the command returned data-0\&. "
.ti -1c
.RI "unsigned char \fBgetCommandResult1\fP (void)"
.br
.RI "In case of command completed without errors, this function returns the command returned data-1\&. "
.ti -1c
.RI "\fBCommandRegisterErrors\fP \fBgetCommandError\fP (void)"
.br
.RI "This function returns the last error condition of an executed command\&. "
.ti -1c
.RI "System::String ^ \fBgetCanCommunicationMonitorString\fP (void)"
.br
.in -1c
.SS "Protected Member Functions"

.in +1c
.ti -1c
.RI "\fBRegister\fP ^ \fBreadCommandRegister\fP (void)"
.br
.RI "This function executes the read of the special device command register\&. "
.ti -1c
.RI "\fBRegister\fP ^ \fBreadErrorRegister\fP (void)"
.br
.RI "This function executes the read of the special device error register\&. "
.ti -1c
.RI "\fBRegister\fP ^ \fBreadStatusRegister\fP (unsigned char index)"
.br
.RI "This function executes the read of the STATUS register of given INDEX\&. "
.ti -1c
.RI "\fBRegister\fP ^ \fBreadDataRegister\fP (unsigned char index)"
.br
.RI "This function executes the read of the DATA register of given INDEX\&. "
.ti -1c
.RI "bool \fBwriteParamRegister\fP (unsigned char index, unsigned char d0, unsigned char d1, unsigned char d2, unsigned char d3)"
.br
.RI "This function executes the write of the PARAMETER(index)\&. "
.ti -1c
.RI "bool \fBwriteParamRegister\fP (unsigned char idx, \fBRegister\fP^ reg)"
.br
.RI "This function executes the write of the PARAMETER(index)\&. "
.ti -1c
.RI "bool \fBwriteDataRegister\fP (unsigned char index, unsigned char d0, unsigned char d1, unsigned char d2, unsigned char d3)"
.br
.RI "This function executes the write of the DATA(index)\&. "
.ti -1c
.RI "bool \fBwriteDataRegister\fP (unsigned char idx, \fBRegister\fP^ reg)"
.br
.RI "This function executes the write of the DATA(index)\&. "
.ti -1c
.RI "\fBCanDeviceProtocol::CanDeviceRegister\fP ^ \fBgetRxRegister\fP (void)"
.br
.RI "This function returns the last read register\&. "
.ti -1c
.RI "virtual void \fBrunningLoop\fP (void)"
.br
.RI "This routine shall be overridden by the subclass module in order to inplement the specific module features\&. "
.ti -1c
.RI "virtual void \fBdemoLoop\fP (void)"
.br
.RI "This routine shall be overridden by the subclass module in order to implement the simulated features\&. "
.ti -1c
.RI "virtual void \fBresetLoop\fP (void)"
.br
.RI "This routine shall be overridden by the subclass module in order to implement the specific operations in case of the remote device reset detection\&. "
.ti -1c
.RI "virtual bool \fBconfigurationLoop\fP (void)"
.br
.RI "This routine shall be overridden by the subclass module in order to implement the specific operations during the device configuration fase\&. "
.ti -1c
.RI "virtual bool \fBsimulCommandNoWaitCompletion\fP (unsigned char code, unsigned char d0, unsigned char d1, unsigned char d2, unsigned char d3, int tmo)"
.br
.RI "This function is called in case of simulation, when the \fBcommandNoWaitCompletion()\fP should be used by the application See the \fBcommandNoWaitCompletion()\fP for details\&. "
.ti -1c
.RI "virtual \fBCanDeviceCommandResult\fP ^ \fBsimulCommandWaitCompletion\fP (unsigned char code, unsigned char d0, unsigned char d1, unsigned char d2, unsigned char d3, int tmo, Object^ src)"
.br
.RI "This function is called in case of simulation, when the \fBcommandWaitCompletion()\fP should be used by the application See the \fBcommandWaitCompletion()\fP for details\&. "
.in -1c
.SS "Protected Attributes"

.in +1c
.ti -1c
.RI "\fBCanDeviceCommunicationMonitor\fP \fBcan_communication_monitor\fP"
.br
.RI "This is the debug class\&. "
.in -1c
.SS "Private Types"

.in +1c
.ti -1c
.RI "enum class \fBCommandRegisterStatus\fP { \fBCommandRegisterStatus::COMMAND_EXECUTING\fP = 1, \fBCommandRegisterStatus::COMMAND_TERMINATED\fP, \fBCommandRegisterStatus::COMMAND_ERROR\fP }"
.br
.RI "Standard constructor\&. "
.ti -1c
.RI "enum class \fBProtocolFrameCode\fP { \fBProtocolFrameCode::FRAME_ERROR\fP = 0, \fBProtocolFrameCode::FRAME_READ_REVISION\fP, \fBProtocolFrameCode::FRAME_READ_ERRORS\fP, \fBProtocolFrameCode::FRAME_READ_COMMAND\fP, \fBProtocolFrameCode::FRAME_READ_STATUS\fP, \fBProtocolFrameCode::FRAME_READ_DATA\fP, \fBProtocolFrameCode::FRAME_READ_PARAM\fP, \fBProtocolFrameCode::FRAME_WRITE_DATA\fP, \fBProtocolFrameCode::FRAME_WRITE_PARAM\fP, \fBProtocolFrameCode::FRAME_STORE_PARAMS\fP, \fBProtocolFrameCode::FRAME_COMMAND_EXEC\fP, \fBProtocolFrameCode::FRAME_DEVICE_RESET\fP }"
.br
.RI "This is the command Code of the protocol frames\&. "
.in -1c
.SS "Private Member Functions"

.in +1c
.ti -1c
.RI "bool \fBsend\fP (unsigned char d0, unsigned char d1, unsigned char d2, unsigned char d3, unsigned char d4, unsigned char d5, unsigned char d6, unsigned char d7, bool bootl)"
.br
.RI "This function sends a frame the the remote device\&. "
.ti -1c
.RI "void \fBthread_can_rx_callback\fP (unsigned short canid, unsigned char *data, unsigned char len)"
.br
.RI "This is the callback to be connected to the CAN reception event\&. "
.ti -1c
.RI "void \fBmainWorker\fP (void)"
.br
.RI "Main Thread routine activated by the Module's Thread\&. "
.ti -1c
.RI "void \fBInternalRunningLoop\fP (void)"
.br
.RI "Internal Loop routine called when the device result configured and correctly running\&. "
.in -1c
.SS "Private Attributes"

.in +1c
.ti -1c
.RI "Thread ^ \fBmain_thread\fP"
.br
.RI "Module internal main thread handle\&. "
.ti -1c
.RI "unsigned short \fBdevice_id\fP"
.br
.RI "Device ID assigned by the Subclass to handle a target remote device\&. "
.ti -1c
.RI "bool \fBsimulator_mode\fP"
.br
.RI "True if the module has been activated in simulation mode\&. "
.ti -1c
.RI "bool \fBrun\fP"
.br
.RI "True if the module has been activated in real mode\&. "
.ti -1c
.RI "bool \fBregister_access_fault\fP"
.br
.RI "True if a series of attempt to send frames to the device should be failed\&. "
.ti -1c
.RI "int \fBregister_access_fault_counter\fP"
.br
.RI "Counter variable to count the consecutive failed attempt to send frame to the remote device\&. "
.ti -1c
.RI "HANDLE \fBrxEvent\fP"
.br
.RI "Event object signaled by the receiving callback\&. "
.ti -1c
.RI "bool \fBrx_pending\fP"
.br
.RI "A reception data is pending\&. "
.ti -1c
.RI "bool \fBdevice_reset\fP"
.br
.RI "A Device reset code has been received\&. "
.ti -1c

PROJECT_MCPU/MCPU/MCPU_MASTER/DEVICES/CanDeviceProtocol.cpp
#include "CanDriver.h"
#include "CanDeviceProtocol.h"
#include "gantry_global_status.h"
#include <thread>
#include <mutex>
#include "Log.h"

using namespace System::Diagnostics;
using namespace System::Collections::Generic;

static std::mutex command_wait_mutex; //!< Mutex used for command with waiting access

CanDeviceProtocol::CanDeviceProtocol(unsigned char devid, LPCWSTR devname) {

    // Gets the handler of the class instance to be used for the Post/Send message functions.
    //hwnd = static_cast<HWND>(Handle.ToPointer());
    device_id = devid;
    device_reset = false;
    rx_register = gcnew CanDeviceProtocol::CanDeviceRegister();
    tx_register = gcnew CanDeviceProtocol::CanDeviceRegister();
    attempt = 0;
    communication_error = false;
    run = false;
    simulator_mode = false;
    command_executing = false;
    
    register_access_fault = 0;
    register_access_fault_counter = 0;

    // Creates the Reception event
    rxEvent = CreateEvent(
        NULL,               // default security attributes
        FALSE,              // Automatic reset after signaling
        FALSE,              // initial state is nonsignaled
        devname             // object name
    );


    // Creates the receiving register
    rx_pending = false;
    rx_sequence = 1;


    // Creates the Startup thread
    main_thread = gcnew Thread(gcnew ThreadStart(this, &CanDeviceProtocol::mainWorker));
    main_thread->Name = "DEVICE" + System::Convert::ToString(device_id);
    main_thread->IsBackground = true; // Important!!! This is necessary to allow the thread to exit when the program exits !!!
    main_thread->Start();

}

void CanDeviceProtocol::thread_can_rx_callback(unsigned short canid, unsigned char* data, unsigned char len) {
    // Be careful!! This call is runing into the program thread pull, out of the class thread.
    bool bootloader = false;

    // No more transaction with a register access fault activated
    if (register_access_fault) return ;
    if (len < 8) return;
    if (!rx_pending) return;


    // Checks the canId address
    if (canid == device_id + 0x140) bootloader = false;
    else if (canid == device_id + 0x100) bootloader = true;
    else
    {
        return; // Not a target
    }

    bool decode_ok = rx_register->decode(data, bootloader);

    // Verifies if there is a Device Reset condition to be signaled
    if ((!bootloader) && (rx_register->b1 == (unsigned char) ProtocolFrameCode::FRAME_DEVICE_RESET)) {
        device_reset = true;
        SetEvent(rxEvent);
        return;
    }

    // Wrong destination address
    if (bootloader != tx_register->bootloader) return;
    
    // Wrong CRC
    if (!decode_ok) {
        LogClass::logInFile("Device Board <" + System::Convert::ToString(device_id) + ">: wrong crc");        
        return;
    }

    // Invalid sequence number
    if(bootloader && (rx_register->b0 != 0xFF) && (rx_register->b0 != tx_register->b0) ) return;
    if (!bootloader && (rx_register->b0 != tx_register->b0)) {
        can_communication_monitor.repeated_messages++;
        return;
    }

    
    // Invalid register access
    if ( (bootloader && (rx_register->b0 == 0xFF)) || (!bootloader && (rx_register->b1 == 0)) ){
        if (register_access_fault_counter > 10) {
            register_access_fault = true;
            LogClass::logInFile("Device Board <" + System::Convert::ToString(device_id) + ">: FATAL!! invalid register access. Stop communication");
        } else {
            register_access_fault_counter++;
            LogClass::logInFile("Device Board <" + System::Convert::ToString(device_id) + ">: invalid register access");
        }

        SetEvent(rxEvent);
        return;
    }

    // Wrong Command match
    if ((!bootloader) && (rx_register->b1 != tx_register->b1)) {
        LogClass::logInFile("Device Board <" + System::Convert::ToString(device_id) + ">: wrong command match");
        return;
    }

    // Checks that the received Object registers matches with the expected    
    rx_pending = false; 
    rxOk = true;

    // Wake the waiting command
    SetEvent(rxEvent);

    return;
}

bool CanDeviceProtocol::send(unsigned char d0, unsigned char d1, unsigned char d2, unsigned char d3, unsigned char d4, unsigned char d5, unsigned char d6, unsigned char d7, bool bootl) {

    unsigned char buffer[8];
    DWORD dwWaitResult;
    unsigned short canid;
    static int count = 0;

    // No more transaction with a register access fault activated
    if (register_access_fault) return false;

    if (bootl) {
        canid = device_id + 0x100;
        tx_register->set(d0, d1, d2, d3, d4, d5, d6, d7, bootl);
        tx_register->format(buffer);
    }
    else {
        canid = device_id + 0x140;
        rx_sequence++;
        if (!rx_sequence) rx_sequence = 1;
        tx_register->set(rx_sequence, d1, d2, d3, d4, d5, d6, d7, bootl);
        tx_register->format(buffer);
        buffer[7] = tx_register->crc;
    }
    
    // Activates the transmission
    rxOk = false;
    can_communication_monitor.sent_messages++;
    long start = clock();
    if (!CanDriver::multithread_send(canid, buffer, 8, simulator_mode))   return false;
    rx_pending = true;

    count++;

    // Waits to be signalled: waits for 100ms
    dwWaitResult = WaitForSingleObject(rxEvent, 100);
    rx_pending = false;

    // Checks if the Event has been signalled or it is a timeout event.
    if ((dwWaitResult != WAIT_OBJECT_0) || (!rxOk)) {
        count = 0;
        can_communication_monitor.unreceived_messages++;
        if (can_communication_monitor.unreceived_messages > 5) communication_error = true;
        return false;
    }
    can_communication_monitor.unreceived_messages = 0;
    communication_error = false;

    long stop = clock();
    can_communication_monitor.txrx_time = ((double)(stop - start)) / (double)CLOCKS_PER_SEC;
    can_communication_monitor.updateStatistics();


    return true;
    
}


void CanDeviceProtocol::mainWorker(void) {
    LogClass::logInFile("Device Board <" + System::Convert::ToString(device_id) + ">: thread started: wait run command");
    while (!run) {
        std::this_thread::sleep_for(std::chrono::milliseconds(100));
    }

    // Add the callback handling the  reception
    CanDriver::canrx_device_event += gcnew CanDriver::delegate_can_rx_frame(this, &CanDeviceProtocol::thread_can_rx_callback);
    internal_status = status_options::WAITING_CAN_DRIVER_CONNECTION;

    // Running Mode activation
    if (simulator_mode) {
        LogClass::logInFile("Device Board <" + System::Convert::ToString(device_id) + ">: Module Run in Simulated mode");
    }
    else {
        LogClass::logInFile("Device Board <" + System::Convert::ToString(device_id) + ">: Module Normal mode");
    }
    
   
    while (1) {
        
        if (!CanDriver::isConnected(simulator_mode)) internal_status = status_options::WAITING_CAN_DRIVER_CONNECTION;
      
        switch (internal_status) {
        
        case status_options::WAITING_CAN_DRIVER_CONNECTION: // Waits for the CAN DRIVER connection
            if (CanDriver::isConnected(simulator_mode)) {
                internal_status = status_options::WAITING_REVISION;
                LogClass::logInFile("Device Board <" + System::Convert::ToString(device_id) + ">: Wait Revision");
            }else{
                std::this_thread::sleep_for(std::chrono::milliseconds(500));
            }    
            
            // When not connected the demo Loop is used to set some standard device status 
            if(simulator_mode) demoLoop();
            break;
        

        case status_options::WAITING_REVISION: // Waits for the device revision acquisition     
            if (send(GET_BOOTLOADER_INFO)) {
                bootloader_status = rx_register->b1;
                boot_maj = rx_register->b2;
                boot_min = rx_register->b3;
                boot_sub = rx_register->b4;
                app_maj = rx_register->b5;
                app_min = rx_register->b6;
                app_sub = rx_register->b7;
                internal_status = status_options::DEVICE_CONFIGURATION;
                LogClass::logInFile("Device Board <" + System::Convert::ToString(device_id) + ">: Configuration");
                break;
            }

            std::this_thread::sleep_for(std::chrono::milliseconds(500));
            
            break;

        case status_options::DEVICE_CONFIGURATION: // Waits for the device configuration
            if (configurationLoop()) {
                LogClass::logInFile("Device Board <" + System::Convert::ToString(device_id) + ">: Running");
                internal_status = status_options::DEVICE_RUNNING;
            }
            std::this_thread::sleep_for(std::chrono::milliseconds(500));            
            break;

        case status_options::DEVICE_RUNNING: // Device running
            // If the device should reset
            if (device_reset) {
                LogClass::logInFile("Device Board <" + System::Convert::ToString(device_id) + ">: Reset Detected");
                device_reset = false;
                internal_status = status_options::WAITING_CAN_DRIVER_CONNECTION;
                resetLoop();
                std::this_thread::sleep_for(std::chrono::milliseconds(500));
                break;
            }
            InternalRunningLoop();
            break;


        } // Switch internal_status
    }

}

/// <summary>
/// This function handles the Module workflow.
/// </summary>
/// 
/// The Module workflow is following described:
/// + run the application loop;
/// + handles the command execution;
/// + if the command takes time to be completed, a nested loop waits for the command completion;
///     + In the nested loop, every 100ms, he application loop is always called;
/// 
/// <param name=""></param>
void CanDeviceProtocol::InternalRunningLoop(void) {

    // Calls the application loop 
    runningLoop();
   
    if (!command_executing) return;

    // handles the command execution  
    if (isCommunicationError()) {
        LogClass::logInFile("Device Board <" + System::Convert::ToString(device_id) + ">: Command Communication error");
        command_error = (System::Byte)CommandRegisterErrors::COMMAND_COMMUNICATION_ERROR;
        command_executing = false;
        return;
    }
       
    int timeout = command_tmo;
    
    // Reads the command register of the remote device
    if (!send(GET_COMMAND_REGISTER)) return;


    if (getRxRegister()->b3 == (System::Byte) CommandRegisterStatus::COMMAND_EXECUTING) {
        LogClass::logInFile("Device Board <" + System::Convert::ToString(device_id) + ">: Command Execution Busy");
        command_error = (System::Byte) CommandRegisterErrors::COMMAND_ERROR_BUSY;
        command_executing = false;
        return;
    }

    // Sends the Requested Command
    if (!send(0, 10, command_code, command_d0, command_d1, command_d2, command_d3, 0, false)) {
        LogClass::logInFile("Device Board <" + System::Convert::ToString(device_id) + ">: Command Not Executed");
        command_error = (System::Byte)CommandRegisterErrors::COMMAND_ERROR_MOMENTARY_DISABLED;
        command_executing = false;
        return;
    }
        
    // Gets the command returning code
    if (getRxRegister()->b3 == (System::Byte)CommandRegisterStatus::COMMAND_ERROR) {
        LogClass::logInFile("Device Board <" + System::Convert::ToString(device_id) + ">: Command Execution Error");
        command_error = getRxRegister()->b6;
        command_executing = false;
        return;
    }

    // Command is immediate
    if (getRxRegister()->b3 == (System::Byte)CommandRegisterStatus::COMMAND_TERMINATED) {
        LogClass::logInFile("Device Board <" + System::Convert::ToString(device_id) + ">: Command Immediate Executed");
        command_error = (System::Byte) CommandRegisterErrors::COMMAND_NO_ERROR;
        command_ris0 = getRxRegister()->b4;
        command_ris1 = getRxRegister()->b5;
        command_executing = false;
        return;
    }

    // If the command takes time to be completed a new loop is handled here
    while (true) {
            
        // During the command loop, the application loop shall continue to work!
        runningLoop();

        timeout--;
        if (!timeout) {
            LogClass::logInFile("Device Board <" + System::Convert::ToString(device_id) + ">: Command Execution Timeout");
            command_error = (System::Byte)CommandRegisterErrors::COMMAND_DEVICE_TMO;
            command_executing = false;
            return;
        }

        while (!send(GET_COMMAND_REGISTER)) {
            if (isCommunicationError()) {
                LogClass::logInFile("Device Board <" + System::Convert::ToString(device_id) + ">: Command Read Error");
                command_error = (System::Byte)CommandRegisterErrors::COMMAND_DEVICE_TMO;
                command_executing = false;
                return;
            }
        }

            
        // The command is still executing
        if (getRxRegister()->b3 == (System::Byte)CommandRegisterStatus::COMMAND_EXECUTING) {
            std::this_thread::sleep_for(std::chrono::milliseconds(100));
            continue;
        }

        if (getRxRegister()->b3 == (System::Byte)CommandRegisterStatus::COMMAND_TERMINATED) {
            LogClass::logInFile("Device Board <" + System::Convert::ToString(device_id) + ">: Command Successfully Terminated");
            command_error = (System::Byte)CommandRegisterErrors::COMMAND_NO_ERROR;
            command_ris0 = getRxRegister()->b4;
            command_ris1 = getRxRegister()->b5;
            command_executing = false;
            return;
        }

        // Error condition
        LogClass::logInFile("Device Board <" + System::Convert::ToString(device_id) + ">: Command Terminated in Error");
        command_error = getRxRegister()->b6;
        command_executing = false;
        return;
            

        // Repeats the read command register
        std::this_thread::sleep_for(std::chrono::milliseconds(100));
    }

    // Command completed
    return;
}


void CanDeviceProtocol::runningLoop(void) {    
    std::this_thread::sleep_for(std::chrono::microseconds(500000));
    return;
}

void CanDeviceProtocol::resetLoop(void) {        
    std::this_thread::sleep_for(std::chrono::microseconds(500000));
    return ;
}
bool CanDeviceProtocol::configurationLoop(void) {    
    std::this_thread::sleep_for(std::chrono::microseconds(500000));
    return true;
}

bool CanDeviceProtocol::commandNoWaitCompletion(unsigned char code, unsigned char d0, unsigned char d1, unsigned char d2, unsigned char d3, int tmo) {
    // Protect the function with a mutex
    const std::lock_guard<std::mutex> lock(command_wait_mutex);

    if ((simulator_mode) && (!CanDriver::isConnected(simulator_mode)))  return simulCommandNoWaitCompletion(code, d0, d1, d2, d3, tmo);

    if (command_executing) return false;
    if (internal_status != status_options::DEVICE_RUNNING) return false;

    command_executing = true;
    command_tmo = tmo;
    command_code = code;
    command_d0 = d0;
    command_d1 = d1;
    command_d2 = d2;
    command_d3 = d3;
    command_error = 0;

    return true;
}

bool CanDeviceProtocol::commandNoWaitCompletion(CanDeviceCommand^ command, int tmo) {
    // Protect the function with a mutex
    const std::lock_guard<std::mutex> lock(command_wait_mutex);
    if (command == nullptr) return false;

    //if ((simulator_mode) && (!CanDriver::isConnected(simulator_mode)))  return simulCommandNoWaitCompletion(command->code, command->d0, command->d1, command->d2, command->d3, tmo);

    if (command_executing) return false;
    if (internal_status != status_options::DEVICE_RUNNING) return false;

    command_executing = true;
    command_tmo = tmo;
    command_code = command->code;
    command_d0 = command->d0;
    command_d1 = command->d1;
    command_d2 = command->d2;
    command_d3 = command->d3;
    command_error = 0;

    return true;
}



CanDeviceProtocol::CanDeviceCommandResult^ CanDeviceProtocol::commandWaitCompletion(unsigned char code, unsigned char d0, unsigned char d1, unsigned char d2, unsigned char d3, int tmo, Object^ src) {
    
    // Protect the function with a mutex
    const std::lock_guard<std::mutex> lock(command_wait_mutex);

    // Verifies that the calling module is different from the class implementing the communication with the device
    if(src == (Object^) this) return gcnew CanDeviceCommandResult(CommandRegisterErrors::COMMAND_INVALID_DEVICE);

    //if ((simulator_mode) && (!CanDriver::isConnected(simulator_mode)))  return simulCommandWaitCompletion(code, d0, d1, d2, d3, tmo, src);


    // Invalid Device Status 
    if (internal_status != status_options::DEVICE_RUNNING) return gcnew CanDeviceCommandResult(CommandRegisterErrors::COMMAND_ERROR_MOMENTARY_DISABLED);

    // Command Busy
    if (command_executing)  return gcnew CanDeviceCommandResult(CommandRegisterErrors::COMMAND_ERROR_BUSY);
    

    // Command preparation
    command_executing = true;
    command_tmo = tmo;
    command_code = code;
    command_d0 = d0;
    command_d1 = d1;
    command_d2 = d2;
    command_d3 = d3;
    command_error = 0;

    // Wait for command completion
    while (!isCommandCompleted()) std::this_thread::sleep_for(std::chrono::milliseconds(100));

    // Returns the result
    return gcnew CanDeviceCommandResult((CommandRegisterErrors) command_error, command_ris0, command_ris1);
    
}

CanDeviceProtocol::CanDeviceCommandResult^ CanDeviceProtocol::commandWaitCompletion(CanDeviceCommand^ command, int tmo, Object^ src) {

    // Protect the function with a mutex
    const std::lock_guard<std::mutex> lock(command_wait_mutex);

    // Verifies that the calling module is different from the class implementing the communication with the device
    if (src == (Object^)this) return gcnew CanDeviceCommandResult(CommandRegisterErrors::COMMAND_INVALID_DEVICE);
    if(command == nullptr) return gcnew CanDeviceCommandResult(CommandRegisterErrors::COMMAND_ERROR_INVALID_PARAM);

    //if ((simulator_mode) && (!CanDriver::isConnected(simulator_mode)))  return simulCommandWaitCompletion(command->code, command->d0, command->d1, command->d2, command->d3, tmo, src);


    // Invalid Device Status 
    if (internal_status != status_options::DEVICE_RUNNING) return gcnew CanDeviceCommandResult(CommandRegisterErrors::COMMAND_ERROR_MOMENTARY_DISABLED);

    // Command Busy
    if (command_executing)  return gcnew CanDeviceCommandResult(CommandRegisterErrors::COMMAND_ERROR_BUSY);


    // Command preparation
    command_executing = true;
    command_tmo = tmo;
    command_code = command->code;
    command_d0 = command->d0;
    command_d1 = command->d1;
    command_d2 = command->d2;
    command_d3 = command->d3;
    command_error = 0;

    // Wait for command completion
    while (!isCommandCompleted()) std::this_thread::sleep_for(std::chrono::milliseconds(100));

    // Returns the result
    return gcnew CanDeviceCommandResult((CommandRegisterErrors)command_error, command_ris0, command_ris1);

}


CanDeviceProtocol::Register^ CanDeviceProtocol::readCommandRegister(void) {
    return nullptr;
}
CanDeviceProtocol::Register^ CanDeviceProtocol::readErrorRegister(void) {
    
    if (!send(
        (System::Byte)0, // Internally set to the sequence number
        (System::Byte) ProtocolFrameCode::FRAME_READ_ERRORS, // Frame code
        (System::Byte)0, // Index not used
        (System::Byte)0, // D0 not used
        (System::Byte)0, // D1 not used
        (System::Byte)0, // D2 not used
        (System::Byte)0, // D3 not used
        (System::Byte)0, // CRC internally set
        false)
        ) return nullptr;

    return gcnew Register(getRxRegister()->b3, getRxRegister()->b4, getRxRegister()->b5, getRxRegister()->b6);

}
CanDeviceProtocol::Register^ CanDeviceProtocol::readStatusRegister(unsigned char index) {

    if (!send(
        (System::Byte) 0, // Internally set to the sequence number
        (System::Byte) ProtocolFrameCode::FRAME_READ_STATUS, // Frame code
        (System::Byte) index, // Status register index
        (System::Byte)0, // d0
        (System::Byte)0, // D1
        (System::Byte)0, // D2 
        (System::Byte)0, // D3
        (System::Byte)0, // CRC internally set
        false)// Bootloader == true, Application = false
    ) return nullptr;

    return gcnew Register(getRxRegister()->b3, getRxRegister()->b4, getRxRegister()->b5, getRxRegister()->b6);
    
}

CanDeviceProtocol::Register^ CanDeviceProtocol::readDataRegister(unsigned char index) {
    return nullptr;
}
bool CanDeviceProtocol::writeParamRegister(unsigned char index, unsigned char d0, unsigned char d1, unsigned char d2, unsigned char d3) {
    return send(
        (System::Byte)0, // Internally set to the sequence number
        (System::Byte)ProtocolFrameCode::FRAME_WRITE_PARAM, // Frame code
        (System::Byte)index, // PARAM register index
        (System::Byte)d0, // d0
        (System::Byte)d1, // D1
        (System::Byte)d2, // D2 
        (System::Byte)d3, // D3
        (System::Byte)0, // CRC internally set
        false); // Bootloader == true, Application = false

}
bool CanDeviceProtocol::writeDataRegister(unsigned char index, unsigned char d0, unsigned char d1, unsigned char d2, unsigned char d3) {

    return send(
        (System::Byte)0, // Internally set to the sequence number
        (System::Byte)ProtocolFrameCode::FRAME_WRITE_DATA, // Frame code
        (System::Byte)index, // DATA register index
        (System::Byte)d0, // d0
        (System::Byte)d1, // D1
        (System::Byte)d2, // D2 
        (System::Byte)d3, // D3
        (System::Byte)0, // CRC internally set
        false); // Bootloader == true, Application = false

}
